tools/first_observation_experiment.rb

```diff
@@ -13,13 +15,135 @@
 
 where [options] are:
 EOS
-  opt :user_data_path, "Path to the user data file.", type: :string, short: "-t"
+  opt :cohort_data_path, "Path to the cohort data file.", type: :string, short: "-t"
+end
+
+def get_obs( day, users )
```

Comment:
This "tool" is very complicated with a handful of methods and then a few hundred lines of basically a "main" method. It doesn't have to happen now, but eventually it would be great to create a class for all of this stuff and break up the code that isn't wrapped in methods now, into smaller logical methods. That way it could be tested, and it would be easier to follow. Then the tool can instantiate the class if it needs instances, or call static methods of the class with readable method names, which would make it easier to understand what the tool is doing.

Reply:
yes - the plan is to experiment with this for a few weeks and if it looks like its succeeding at driving retention or yielding useful data towards that goal, then move this tool into the main codebase and also replace the CSV file with some database tables.